### ISSUE TITLE
Discutils.Ntfs: equivalent of OpenFileById(), all clusters...

### ISSUE THREAD

**Issue by jol64** [DiscUtils/DiscUtils#115]
I need the equivalent of OpenFileById() which I did not find in DiscUtils... I guess all of you who ever tried to enumerate the NTFS journal or MFT know why..
Q: am I blind? Is there a method that I overlooked?

I tried the following added method:

        public String GetPath (long index)
        {
            File file = GetFile(index & 0x0000ffffffffffff); 
            if (file == null || file.Names.Count == 0) return null;
            foreach (String s in file.Names)
                if (s.Contains("~") == false) return s;
            return file.Names[0];
         }

Q: I tried BestName, but that is local only and not returning the path - which I need to call other API methods.. as you can see I tried to avoid the short names, but I have seen situations where only that was available. Any suggestions on how to improve?
Q: should one validate the upper 16 bits? how?
Q: any other suggestions?

Now I also wanted a means to obtain all allocated clusters regardless of the alternate stream:

        public Range<long, long>[] PathToClustersAllAttributes(string path)
        {
            string plainPath;
            string attributeName;
            SplitPath(path, out plainPath, out attributeName);

            DirectoryEntry dirEntry = GetDirectoryEntry(plainPath);
            if (dirEntry == null) // || dirEntry.IsDirectory)
            {
                throw new FileNotFoundException("No such file or directory", path);
            }
            File file = GetFile(dirEntry.Reference);

            Range<long, long>[] resultranges = null;
            foreach (var attr in file.AllAttributes)
                if (attr.IsNonResident)
                {
                    var attrranges = attr.GetClusters();
                    if (resultranges == null) resultranges = attrranges;
                    else
                    { // need to merge}
                        Range<long, long>[] mergedesultranges = new Range<long, long>[resultranges.Length + attrranges.Length];
                        Array.Copy(resultranges, 0, mergedesultranges, 0, resultranges.Length);
                        Array.Copy(attrranges, 0, mergedesultranges, resultranges.Length, attrranges.Length);
                        resultranges = mergedesultranges;
                    }
                }
            return resultranges;
        }

works. Now obviously I first need to call GetPath() and then PathToClustersAllAttributes(). Similar call pattern with other calls. Of course I can also add PathToClustersAllAttributes(long index), but I am wondering whether there is a better approach than to duplicate essentially all API methods.

One approach could be to introduce a NtfsFileIdentifier that has two constructors, one with a path, one with an index, other options could be also the object id. Then change all methods to accept a NtfsFileIdentifier. NtfsFileIdentifier could then know how to resolve to a DirectoryEntry and (for readonly) even cache that.

Another approach could be to implement some caching on DirectoryEntry and integrate GetPath with that..

Thoughts?

**Comment by LordMike** [comment]
> Q: am I blind? Is there a method that I overlooked?
So. I've created another [NTFS library](https://github.com/lordmike/ntfslib) which should be able to provide you with nitty-gritty access to structures on disk - I'm not currently maintaining it. I'm unsure if DiscUtils exposes these kinds of access to internal structures.

> Q: I tried BestName, but that is local only and not returning the path - which I need to call other API methods.. as you can see I tried to avoid the short names, but I have seen situations where only that was available. Any suggestions on how to improve?

Once you have access to the structures, you'll get access to all attributes in a given file record (remember external attributes). Of those, you'll want the FILENAME attributes, and then filtering by FileNamespace. The Namespace will be DOS names (8.3 names), Posix, Win32 ("Regular" names) and Win32AndDos (Regular names that also work in DOS). I'd go for them in this order: Win32, Win32AndDos, Posix and then DOS.

A file can have multiple FILENAMES (Hardlinks).

> Q: should one validate the upper 16 bits? how?
The File Reference should be a 48/16 bit or 32/32 (I don't remember) structure, of which the first part is the actual reference and the last part is the "version" of the reference. At least I remember it as the first part being the id... Are you sure your code is correct?

> Q: any other suggestions?
Building paths is slightly tricky. Building _a_ path is easy, building _all possible_ paths can be impossible.

> Now I also wanted a means to obtain all allocated clusters regardless of the alternate stream .. works. Now obviously I first need to call GetPath() and then PathToClustersAllAttributes(). Similar call pattern with other calls. Of course I can also add PathToClustersAllAttributes(long index), but I am wondering whether there is a better approach than to duplicate essentially all API methods.

This is fun. My other library also exposes the DATA attributes, which you'll want. They have sequences of clusters that make up a given file. It should all be parsed and readily accessible in that library. I'm unsure if DiscUtils exposes the same records. Note that some attributes will reference Resident data - in those cases there will be no clusters.

**Comment by jol64** [comment]
> So. I've created another [NTFS library](https://github.com/lordmike/ntfslib) which should be able to provide you with nitty-gritty access to structures on disk - I'm not currently maintaining it. I'm unsure if DiscUtils exposes these kinds of access to internal structures.

One of the nice and sometimes also difficult aspects of OSS is that there are so much code to choose from. Actually my impression was that DiscUtils is more popular, and I also noticed you kind of support both!

In fact I might switch to any library or incorporate the relevant logic into my own code. My first goal is to get the code running, and then see what fits best.

> A file can have multiple FILENAMES (Hardlinks).

> Building paths is slightly tricky. Building _a_ path is easy, building _all possible_ paths can be impossible

Sure. In fact any of these should do as it ultimately resolves to the same index (and clusters) again. However I prefer the one that is most readable - simplifies support. DiscUtils already gives me Names, but not the most readable one. Also I suspect if I use the FILENAME attribute, I have to construct the path myself.
As usual there are many options, just looking for kind of elegance..

> Of course you're correct. 

I learned the hard way that´s not always true.

> Your code has probably run, ..

yes, except for the 48bit logic was somewhere else. But the fact that it worked in my special and read-only scenario is not a hint that validation is unnecessary for a general solution. That´s also why I preferred the issue/discussion first over suggesting a pull request.. which obviously also depends on whether someone else considers that useful...

> This is fun. My other library also exposes the DATA attributes, which you'll want. They have sequences of clusters that make up a given file. It should all be parsed and readily accessible in that library. I'm unsure if DiscUtils exposes the same records. Note that some attributes will reference Resident data - in those cases there will be no clusters.

Not just $DATA, any non-resident attribute.

**Comment by LordMike** [comment]
Ah. You're making a filesystem extraction tool? 

I'll recommend taking special care of the first 24 records (ignore the bad sectors f.ex.), and then simply doing one pass of all records and taking all those marked as `In Use`. For my initial program doing this, I started doing recursive parsing of f.ex. non-resident attribute lists to locate extra clusters in use, finding their attributes and so forth - but the reality is that a single pass of all records and then "marking" their non-resident data is enough.

For my case, I didn't need the DATA, I just needed the metadata for offsite parsing.

**Comment by EricZimmerman** [comment]
good discussions. Not to highjack anything, but I wanted to mention that i wrote an offline, forensically motivated MFT parser that can resolve paths and supports all the structures which you may find interesting. you can find it at https://github.com/EricZimmerman/MFT. There is sure to be some overlap with https://github.com/LordMike/NtfsLib as well.

I have a method in there that builds and then pretty much instantly returns full paths, including hardlinks and all that as well that may prove useful (GetFullParentPath) via the entry and sequence number, which every FILENAME attribute will have.  This code was a pain to get right in an efficient way and I threw away a lot of code getting there!

it also handles deleted file recovery, including rebuilding paths, taking into account how FILE records behave on deletion, with seq #s increasing, etc. there are helpers for IsADS, HasADS, Copied, etc. should those things be of use. it also knows how to get the "right" Filename that most people want by default, as well as the size, and so on.

it also is one of only 2 implementations that i know of that does not suffer from finding all related data when a non-resident attribute list is in play. I do not do any kind of disk reads and do all the parsing with the MFT file only and will find and relate all the extension records to the base records, which then gets you the complete view of files.

from a speed perspective, it can process and dump multi-gigabyte MFTs in a few minutes. Most "average" size MFTs are done in a few seconds.

from the sounds of it my MFT program will, given only the MFT bytes, build collections and many things that will make your life easier for you, at least based on the descriptions above.

i have a cli front end on it you can find at https://github.com/EricZimmerman/MFTECmd and have a few posts about it at https://binaryforay.blogspot.com/2018/06/introducing-mftecmd.html

**Comment by EricZimmerman** [comment]
i love me some NTFS, so these kinds of discussions are great!

**Comment by jol64** [comment]
> Ah. You're making a filesystem extraction tool?

> from a speed perspective, it can process and dump multi-gigabyte MFTs in a few minutes. Most "average" size MFTs are done in a few seconds.

No. I am writing backup software (and not OSS, but free trial). Thus I need to detect changed files with good performance...

> I'll recommend taking special care of the first 24 records

Thanks, I am aware of that these are not entirely covered by USN records..
